PancakeSwap-Bots/Large-Delegators-Vote-Change/src/agent.ts

```diff
@@ -0,0 +1,85 @@
+import {
+  ethers,
+  Finding,
+  FindingSeverity,
+  getEthersProvider,
+  HandleTransaction,
+  Initialize,
+  LogDescription,
+  TransactionEvent,
+} from "forta-agent";
+import { NetworkManager } from "forta-agent-tools";
+import BN from "bignumber.js";
+import { DELEGATE_VOTES_CHANGED_EVENT } from "./abi";
+import { DATA, NetworkData } from "./config";
+import { createFinding } from "./findings";
+import {
+  ABSOLUTE_THRESHOLD,
+  HIGH_THRESHOLD,
+  LOW_THRESHOLD,
+  MEDIUM_THRESHOLD,
+  MIN_PREVIOUS_BALANCE,
+} from "./thresholds";
+
+const networkManager: NetworkManager<NetworkData> = new NetworkManager(DATA);
+
+const provideInitialize = (
+  networkManager: NetworkManager<NetworkData>,
+  provider: ethers.providers.Provider
+): Initialize => {
+  return async () => {
+    BN.set({ DECIMAL_PLACES: 5 });
+    await networkManager.init(provider);
+  };
+};
+
+const provideHandleTransaction = (networkManager: NetworkManager<NetworkData>): HandleTransaction => {
+  return async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    const contractAddress: string = networkManager.get("cakeAddress");
+
+    // filter the transaction logs for events
+    const delegateVotesChangedEvents: LogDescription[] = txEvent.filterLog(
+      DELEGATE_VOTES_CHANGED_EVENT,
+      contractAddress
+    );
+
+    delegateVotesChangedEvents.forEach((delegateVotesChangedEvent) => {
+      // extract event arguments
+      const { delegate, previousBalance, newBalance } = delegateVotesChangedEvent.args;
+
+      let metadata = {
+        delegate,
+        previousBalance: previousBalance.toString(),
+        newBalance: newBalance.toString(),
+      };
+
+      let delta: BN = new BN(newBalance.toString()).minus(previousBalance.toString()); // difference between balances
+
+      if (MIN_PREVIOUS_BALANCE.lt(previousBalance.toString())) {
+        let deltaPercentage: BN = delta.dividedBy(new BN(previousBalance.toString())); // previousBalance.div(DECIMALS).toNumber(); //percentage of vote increase
+
+        let description: string = deltaPercentage.multipliedBy(100).toString() + " %";
+
+        if (deltaPercentage.gte(HIGH_THRESHOLD)) {
+          findings.push(createFinding(description, metadata, FindingSeverity.High));
+        } else if (deltaPercentage.gte(MEDIUM_THRESHOLD)) {
+          findings.push(createFinding(description, metadata, FindingSeverity.Medium));
+        } else if (deltaPercentage.gte(LOW_THRESHOLD)) {
+          findings.push(createFinding(description, metadata, FindingSeverity.Low));
+        }
+      } else if (previousBalance.eq(0) && ABSOLUTE_THRESHOLD.lte(newBalance.toString())) {
```

Comment:
why does the previousBalance have to be zero in order to emit this alert? what happens if the previous balance is 100 and new balance is 11000000000000000000? is this case needed to be omitted for some reason?

Reply:
So in the thresholds file we have a minimum previous balance to create a finding, it could be 100, and if the percentage increase exceeds any of the thresholds it creates a finding. But in case the previous balance is 0 we can't calculate percentages so I made a special condition for it, if the new balance exceeds a certain threshold it creates a finding

Reply:
yes, percentage approach and absolute approach can be used separately or simultaneously